### PR TITLE
Feature/kg 643

### PR DESCRIPTION
**Pull request by DavidIzaac**

### CODE REVIEW

File: console/src/app/maps/map-detail/map-detail.component.ts [hunk 1]
```diff
@@ -252,12 +252,19 @@ export class MapDetailComponent implements OnInit, OnDestroy {
     this.downloadJson = this.sanitizer.bypassSecurityTrustUrl('data:text/json;charset=UTF-8,' + encodeURIComponent(JSON.stringify(structure)));
   }
 
+  getMapResultIdAsObservable(){
```
Comment (Or-Sharbat) [inline console/src/app/maps/map-detail/map-detail.component.ts]: Where you use this function?

File: console/src/app/maps/map-detail/map-result/map-result.component.ts [hunk 2]
```diff
@@ -144,6 +149,61 @@ export class MapResultComponent implements OnInit, OnDestroy {
     this.popupService.openComponent(RawOutputComponent,{messages:messages})
   }
 
+  buildExecutionFromSocket(selectedExecution,result){ 
+
```
Comment (Or-Sharbat) [inline console/src/app/maps/map-detail/map-result/map-result.component.ts]: add some documentation

File: server/api/services/map-execution.service.js [hunk 3]
```diff
@@ -839,6 +875,22 @@ function runProcess(map, structure, runId, agent, process) {
         }
 
         if (!passProcessCondition(runId, agent, process)) {
+            let res = {
+                process:{
```
Comment (Or-Sharbat) [inline server/api/services/map-execution.service.js]: Better to put this code in a new function.

File: server/api/services/map-execution.service.js [hunk 4]
```diff
@@ -1075,7 +1128,7 @@ async function executeAction(map, structure, runId, agent, process, processIndex
      * @returns {object} - action result
      */
     function runAction() {
-
+        let actionResult = actionData;
```
Comment (Or-Sharbat) [inline server/api/services/map-execution.service.js]: you dont use action result?

File: server/api/services/map-execution.service.js [hunk 5]
```diff
@@ -1093,20 +1146,47 @@ async function executeAction(map, structure, runId, agent, process, processIndex
...
+                updateActionContext(runId, agent.key, process.uuid, processIndex, action, actionResult)
                 return runAction();
             }
-            actionData.finishTime = new Date();
+            actionResult.finishTime = new Date();
             let msg = `'${process.name} ' - '${action.name}' result: ${JSON.stringify(result)} (${agent.name})`
             _updateRawOutput(map._id, runId, msg, result.status)
-            updateActionContext(runId, agent.key, process.uuid, processIndex, action, actionData);
+            updateActionContext(runId, agent.key, process.uuid, processIndex, action, actionResult);
+            let res = {
```
Comment (Or-Sharbat) [inline server/api/services/map-execution.service.js]: put in a new function. better in "updateActionContext"

File: console/src/app/maps/map-detail/map-result/map-result.component.ts [hunk 6]
```diff
@@ -157,28 +217,56 @@ export class MapResultComponent implements OnInit, OnDestroy {
...
+                this.setActionToSelectedExecution(selectedExecution,action)
+              })  
+            
+          });
+
+          this.ongoingExecutionSocket.on('updateFinishTime',(data) => {
+            this.selectedExecution.finishTime = data.execution;
+            if(data.process){
+              let processes = [];
+              this.selectedExecution.agentsResults.forEach(agent => {
```
Comment (kadosh1000) [inline console/src/app/maps/map-detail/map-result/map-result.component.ts]: @DavidIzaac change this to use a simple for loop, running over the processes and add the finishTime to the relevant process and break;

File: console/src/app/maps/interfaces/process-list.interface.ts [hunk 7]
```diff
@@ -2,5 +2,5 @@ export interface IProcessList {
     name: string;
     index: number;
     overall: number;
-    uuid: string;
+    process: string;
```
Comment (kadosh1000) [inline console/src/app/maps/interfaces/process-list.interface.ts]: @DavidIzaac revert back to uuid, this change caused a lot of places that do `process.process` which is not correct and misleading

File: console/src/app/maps/map-detail/map-detail.component.ts (hunk 1 above)
Comment (kadosh1000) [inline console/src/app/maps/map-detail/map-detail.component.ts]: @DavidIzaac if this function is not in use, please remove it. 
What is the `mapResultId` used for? where is it being used?

File: server/api/services/map-execution.service.js [hunk 8]
```diff
@@ -744,6 +768,14 @@ function updateAgentContext(runId, agent, agentData) {
     dbUpdates.updateAgent(options)
 }
 
+function sendFinishTimeToClient(runId,data){
+    let nsp = socketService.getNamespaceSocket('execution-update-'+executions[runId].mapResultId);
+    Object.keys(nsp.sockets).forEach(socket => {
```
Comment (kadosh1000) [inline server/api/services/map-execution.service.js]: @DavidIzaac why not using `nsp.emit` to send to all sockets in the namespace?
https://socket.io/docs/rooms-and-namespaces/

File: server/api/services/map-execution.service.js (hunk 4 above)
Comment (kadosh1000) [inline server/api/services/map-execution.service.js]: @DavidIzaac what is this for? why not just using actionData directly?

File: server/api/services/map-execution.service.js (hunk 5 above)
Comment (kadosh1000) [inline server/api/services/map-execution.service.js]: @DavidIzaac move this to a function. call the function `sendActionResultToClients`

**Pull request closed without merging**